File: src-tauri/src/db/project.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use super::Database;

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Project {
    pub id: String,
    #[serde(rename = "organizationId")]
    #[sqlx(rename = "organizationId")]
    pub organization_id: String,
    pub name: String,
    #[serde(rename = "clientId")]
    #[sqlx(rename = "clientId")]
    pub client_id: Option<String>,
    #[serde(rename = "clientName")]
    #[sqlx(rename = "clientName")]
    pub client_name: Option<String>,
    #[serde(rename = "startDate")]
    #[sqlx(rename = "startDate")]
    pub start_date: Option<i64>,
    #[serde(rename = "endDate")]
    #[sqlx(rename = "endDate")]
    pub end_date: Option<i64>,
    #[serde(rename = "archivedAt")]
    #[sqlx(rename = "archivedAt")]
    pub archived_at: Option<i64>,
    #[serde(rename = "createdAt")]
    #[sqlx(rename = "createdAt")]
    pub created_at: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateProjectRequest {
    pub id: String,
    #[serde(rename = "organizationId")]
    pub organization_id: String,
    pub name: String,
    #[serde(rename = "clientId")]
    pub client_id: Option<String>,
    #[serde(rename = "startDate")]
    pub start_date: Option<i64>,
    #[serde(rename = "endDate")]
    pub end_date: Option<i64>,
    #[serde(rename = "archivedAt")]
    pub archived_at: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateProjectRequest {
    pub name: Option<String>,
    #[serde(rename = "clientId")]
    pub client_id: Option<String>,
    #[serde(rename = "startDate")]
    pub start_date: Option<i64>,
    #[serde(rename = "endDate")]
    pub end_date: Option<i64>,
    #[serde(rename = "archivedAt")]
    pub archived_at: Option<i64>,
}

impl Database {
    // Project methods
    pub async fn get_projects(&self, organization_id: &str) -> Result<Vec<Project>, sqlx::Error> {
        let projects = sqlx::query_as::<_, Project>(
            "SELECT p.id, p.organizationId, p.name, p.clientId, c.name as clientName, p.startDate, p.endDate, p.archivedAt, p.createdAt 
             FROM projects p 
             LEFT JOIN clients c ON p.clientId = c.id 
             WHERE p.organizationId = ? 
             ORDER BY p.name ASC"
        )
        .bind(organization_id)
        .fetch_all(&self.pool)
        .await?;

        Ok(projects)
    }

    pub async fn get_project(&self, project_id: &str) -> Result<Option<Project>, sqlx::Error> {
        let project = sqlx::query_as::<_, Project>(
            "SELECT * FROM projects WHERE id = ?"
        )
        .bind(project_id)
        .fetch_optional(&self.pool)
        .await?;

        Ok(project)
    }

    pub async fn create_project(&self, project: CreateProjectRequest) -> Result<Project, sqlx::Error> {
        sqlx::query(
            r#"
            INSERT INTO projects (id, organizationId, name, clientId, startDate, endDate, archivedAt)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            "#,
        )
        .bind(&project.id)
        .bind(&project.organization_id)
        .bind(&project.name)
        .bind(&project.client_id)
        .bind(&project.start_date)
        .bind(&project.end_date)
        .bind(&project.archived_at)
        .execute(&self.pool)
        .await?;

        let created_project = self.get_project(&project.id).await?;
        Ok(created_project.unwrap())
    }

    pub async fn update_project(&self, project_id: &str, updates: UpdateProjectRequest) -> Result<Project, sqlx::Error> {
        // Use individual queries for each field to avoid complex dynamic binding
        if let Some(name) = &updates.name {
            sqlx::query("UPDATE projects SET name = ? WHERE id = ?")
                .bind(name)
                .bind(project_id)
                .execute(&self.pool)
                .await?;
        }
        if updates.client_id.is_some() {
            sqlx::query("UPDATE projects SET clientId = ? WHERE id = ?")
                .bind(&updates.client_id)
                .bind(project_id)
                .execute(&self.pool)
                .await?;
        }
        if updates.start_date.is_some() {
            sqlx::query("UPDATE projects SET startDate = ? WHERE id = ?")
                .bind(&updates.start_date)
                .bind(project_id)
                .execute(&self.pool)
                .await?;
        }
        if updates.end_date.is_some() {
            sqlx::query("UPDATE projects SET endDate = ? WHERE id = ?")
                .bind(&updates.end_date)
                .bind(project_id)
                .execute(&self.pool)
                .await?;
        }
        if updates.archived_at.is_some() {
            sqlx::query("UPDATE projects SET archivedAt = ? WHERE id = ?")
                .bind(&updates.archived_at)
                .bind(project_id)
                .execute(&self.pool)
                .await?;
        }

        let updated_project = self.get_project(project_id).await?;
        Ok(updated_project.unwrap())
    }
}

File: db/project.go
package db

import "fmt"

// Project mirrors the projects table (with an optional joined clientName).
type Project struct {
	ID             string  `db:"id"             json:"id"`
	OrganizationID string  `db:"organizationId" json:"organizationId"`
	Name           string  `db:"name"           json:"name"`
	ClientID       *string `db:"clientId"       json:"clientId"`
	ClientName     *string `db:"clientName"     json:"clientName"`
	StartDate      *int64  `db:"startDate"      json:"startDate"`
	EndDate        *int64  `db:"endDate"        json:"endDate"`
	ArchivedAt     *int64  `db:"archivedAt"     json:"archivedAt"`
	CreatedAt      *string `db:"createdAt"      json:"createdAt"`
}

// CreateProjectRequest is the payload for creating a project.
type CreateProjectRequest struct {
	ID             string  `json:"id"`
	OrganizationID string  `json:"organizationId"`
	Name           string  `json:"name"`
	ClientID       *string `json:"clientId"`
	StartDate      *int64  `json:"startDate"`
	EndDate        *int64  `json:"endDate"`
	ArchivedAt     *int64  `json:"archivedAt"`
}

// UpdateProjectRequest is the payload for updating a project.
type UpdateProjectRequest struct {
	Name       *string `json:"name"`
	ClientID   *string `json:"clientId"`
	StartDate  *int64  `json:"startDate"`
	EndDate    *int64  `json:"endDate"`
	ArchivedAt *int64  `json:"archivedAt"`
}

func (d *Database) GetProjects(organizationID string) ([]Project, error) {
	projects := []Project{}
	err := d.DB.Select(&projects, `
		SELECT p.id, p.organizationId, p.name, p.clientId, c.name AS clientName,
		       p.startDate, p.endDate, p.archivedAt, p.createdAt
		FROM projects p
		LEFT JOIN clients c ON p.clientId = c.id
		WHERE p.organizationId = ?
		ORDER BY p.name ASC`,
		organizationID,
	)
	if err != nil {
		return nil, fmt.Errorf("get_projects: %w", err)
	}
	return projects, nil
}

func (d *Database) GetProject(projectID string) (*Project, error) {
	var project Project
	err := d.DB.Get(&project, `
		SELECT p.id, p.organizationId, p.name, p.clientId, c.name AS clientName,
		       p.startDate, p.endDate, p.archivedAt, p.createdAt
		FROM projects p
		LEFT JOIN clients c ON p.clientId = c.id
		WHERE p.id = ?`,
		projectID,
	)
	if err != nil {
		return nil, fmt.Errorf("get_project: %w", err)
	}
	return &project, nil
}

func (d *Database) CreateProject(req CreateProjectRequest) (*Project, error) {
	_, err := d.DB.Exec(`
		INSERT INTO projects (id, organizationId, name, clientId, startDate, endDate, archivedAt)
		VALUES (?, ?, ?, ?, ?, ?, ?)`,
		req.ID, req.OrganizationID, req.Name, req.ClientID,
		req.StartDate, req.EndDate, req.ArchivedAt,
	)
	if err != nil {
		return nil, fmt.Errorf("create_project: %w", err)
	}
	return d.GetProject(req.ID)
}

func (d *Database) DeleteProject(projectID string) (bool, error) {
	tx, err := d.DB.Beginx()
	if err != nil {
		return false, fmt.Errorf("delete_project begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	// projectId has no DB-level FK on timeEntries (migration 0015); null it out manually.
	if _, err = tx.Exec(`UPDATE timeEntries SET projectId = NULL WHERE projectId = ?`, projectID); err != nil {
		return false, fmt.Errorf("delete_project nullify_entries: %w", err)
	}

	res, err := tx.Exec(`DELETE FROM projects WHERE id = ?`, projectID)
	if err != nil {
		return false, fmt.Errorf("delete_project: %w", err)
	}

	if err := tx.Commit(); err != nil {
		return false, fmt.Errorf("delete_project commit: %w", err)
	}

	n, _ := res.RowsAffected()
	return n > 0, nil
}

func (d *Database) UpdateProject(projectID string, updates UpdateProjectRequest) (*Project, error) {
	if updates.Name != nil {
		if _, err := d.DB.Exec(`UPDATE projects SET name = ? WHERE id = ?`, updates.Name, projectID); err != nil {
			return nil, fmt.Errorf("update_project name: %w", err)
		}
	}
	if updates.ClientID != nil {
		if _, err := d.DB.Exec(`UPDATE projects SET clientId = ? WHERE id = ?`, updates.ClientID, projectID); err != nil {
			return nil, fmt.Errorf("update_project clientId: %w", err)
		}
	}
	if updates.StartDate != nil {
		if _, err := d.DB.Exec(`UPDATE projects SET startDate = ? WHERE id = ?`, updates.StartDate, projectID); err != nil {
			return nil, fmt.Errorf("update_project startDate: %w", err)
		}
	}
	if updates.EndDate != nil {
		if _, err := d.DB.Exec(`UPDATE projects SET endDate = ? WHERE id = ?`, updates.EndDate, projectID); err != nil {
			return nil, fmt.Errorf("update_project endDate: %w", err)
		}
	}
	if updates.ArchivedAt != nil {
		if _, err := d.DB.Exec(`UPDATE projects SET archivedAt = ? WHERE id = ?`, updates.ArchivedAt, projectID); err != nil {
			return nil, fmt.Errorf("update_project archivedAt: %w", err)
		}
	}
	return d.GetProject(projectID)
}
